src/library/scala/collection/immutable/TreeMap.scala

```diff
@@ -47,11 +47,39 @@ final class TreeMap[K, +V] private (tree: RB.Tree[K, V])(implicit val ordering:
 
   def get(key: K): Option[V] = RB.get(tree, key)
 
-  def remove(key: K): TreeMap[K,V] =
-    if (!RB.contains(tree, key)) this
-    else new TreeMap(RB.delete(tree, key))
+  def remove(key: K): TreeMap[K,V] = {
+    val t = RB.delete(tree, key)
+    if(t eq tree) this else new TreeMap(t)
```

Comment:
Extract this to a helper `newTreeMapOrSelf`.

I think there are a few more places it could be used: `rangeImpl` `range`.